File: demo/ruby/g_set.rb
#!/usr/bin/ruby

# A CRDT-based grow-only set.

require_relative 'node.rb'
require 'set'

class GSetNode
  def initialize
    @node = Node.new
    @set = Set.new

    @node.on "read" do |msg|
      @node.reply! msg, {type: "read_ok", value: @set.to_a}
    end

    @node.on "add" do |msg|
      element = msg[:body][:element]
      @set.add element
      @node.reply! msg, {type: "add_ok"}
    end

    # Accept a single element from another node
    @node.on "replicate_one" do |msg|
      @set.add msg[:body][:element]
    end

    # Accept an entire value from another node
    @node.on "replicate_full" do |msg|
      @set |= msg[:body][:value]
    end

    # Periodically replicate entire state
    @node.every 5 do
      STDERR.puts "Replicating!"
      @node.other_node_ids.each do |node|
        @node.send! node, ({type: "replicate_full", value: @set.to_a})
      end
    end
  end

  def main!
    @node.main!
  end
end

GSetNode.new.main!

File: demo/rust/src/bin/g_set.rs
/// ```bash
/// $ cargo build
/// $ RUST_LOG=debug ~/Projects/maelstrom/maelstrom test -w g-set --bin ./target/debug/g_set --node-count 2 --concurrency 2n --time-limit 20 --rate 10 --log-stderr
/// ```
use async_trait::async_trait;
use log::debug;
use maelstrom::protocol::Message;
use maelstrom::{done, Node, Result, Runtime};
use serde::{Deserialize, Serialize};
use std::collections::HashSet;
use std::sync::{Arc, Mutex, MutexGuard};
use std::time::Duration;

pub(crate) fn main() -> Result<()> {
    Runtime::init(try_main())
}

async fn try_main() -> Result<()> {
    let runtime = Runtime::new();
    let handler = Arc::new(Handler::default());
    runtime.with_handler(handler).run().await
}

#[derive(Clone, Default)]
struct Handler {
    s: Arc<Mutex<HashSet<i64>>>,
}

#[async_trait]
impl Node for Handler {
    async fn process(&self, runtime: Runtime, req: Message) -> Result<()> {
        let msg: Result<Request> = req.body.as_obj();
        match msg {
            Ok(Request::Read {}) => {
                let data = to_seq(&self.s.lock().unwrap());
                return runtime.reply(req, Request::ReadOk { value: data }).await;
            }
            Ok(Request::Add { element }) => {
                self.s.lock().unwrap().insert(element);
                return runtime.reply(req, Request::AddOk {}).await;
            }
            Ok(Request::ReplicateOne { element }) => {
                self.s.lock().unwrap().insert(element);
                return Ok(());
            }
            Ok(Request::ReplicateFull { value }) => {
                let mut s = self.s.lock().unwrap();
                for v in value {
                    s.insert(v);
                }
                return Ok(());
            }
            Ok(Request::Init {}) => {
                // spawn into tokio (instead of runtime) to not to wait
                // until it is completed, as it will never be.
                let (r0, h0) = (runtime.clone(), self.clone());
                tokio::spawn(async move {
                    loop {
                        tokio::time::sleep(Duration::from_secs(5)).await;
                        debug!("emit replication signal");
                        let s = h0.s.lock().unwrap();
                        for n in r0.neighbours() {
                            let msg = Request::ReplicateFull { value: to_seq(&s) };
                            drop(r0.send_async(n, msg));
                        }
                    }
                });
                return Ok(());
            }
            _ => done(runtime, req),
        }
    }
}

fn to_seq(s: &MutexGuard<HashSet<i64>>) -> Vec<i64> {
    s.iter().copied().collect()
}

#[derive(Serialize, Deserialize)]
#[serde(rename_all = "snake_case", tag = "type")]
enum Request {
    Init {},
    Read {},
    ReadOk { value: Vec<i64> },
    Add { element: i64 },
    AddOk {},
    ReplicateOne { element: i64 },
    ReplicateFull { value: Vec<i64> },
}
